### PR TITLE
feat(bundle): support autoExternal configuration

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

support automatically externalize project dependencies and peerDependencies and not package them into the final bundle.

```
type AutoExternal =
  | boolean
  | {
      dependencies?: boolean;
      devDependencies?: boolean;
      peerDependencies?: boolean;
    };
```

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/config.ts [hunk 1]
```diff
@@ -394,17 +399,24 @@ async function composeLibRsbuildConfig(
...
   const { config: autoExtensionConfig, dtsExtension } =
-    composeAutoExtensionConfig(format!, dirname(configPath), autoExtension);
+    composeAutoExtensionConfig(format!, autoExtension, pkgJson);
   const bundleConfig = composeBundleConfig(config.bundle);
   const targetConfig = composeTargetConfig(config.output?.target);
   const syntaxConfig = composeSyntaxConfig(
     config.output?.syntax,
     config.output?.target,
   );
+  const autoExternalConfig = composeAutoExternal({
```
Comment (Timeless0911) [inline packages/core/src/config.ts:415]: `composeAutoExternalConfig`?

**Comment by Timeless0911** [comment]
Can we update examples `externals` config?

File: packages/core/src/utils/external.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import type { RsbuildConfig } from '@rsbuild/core';
+import type { PkgJson } from '../types';
+import type { AutoExternal } from '../types/config';
+import { logger } from './logger';
+
+export const composeAutoExternalConfig = (options: {
```
Comment (fi3ework) [inline packages/core/src/utils/external.ts:6]: Why not put it in `config.ts`?

File: packages/core/src/utils/external.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    devDependencies: false,
+    ...(autoExternal === true ? {} : autoExternal),
+  };
+
+  // User externals configuration has higher priority than autoExternal
+  // eg: autoExternal: ['react'], user: output: { externals: { react: 'react-1' } }
+  // Only handle the case where the externals type is object, string / string[] does not need to be processed, other types are too complex.
+  const userExternalKeys =
+    userExternals &&
+    Object.prototype.toString.call(userExternals) === '[object Object]'
```
Comment (fi3ework) [inline packages/core/src/utils/external.ts:36]: Does `userExternals instanceof Object` meets the requirement.

File: packages/core/src/utils/external.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      ? Object.keys(userExternals)
+      : [];
+
+  const externals = (
+    ['dependencies', 'peerDependencies', 'devDependencies'] as const
+  )
+    .reduce<string[]>((prev, type) => {
+      if (externalOptions[type]) {
+        return pkgJson[type] ? prev.concat(Object.keys(pkgJson[type]!)) : prev;
```
Comment (fi3ework) [inline packages/core/src/utils/external.ts:45]: 1. No deduplication in `prev`.
2. If a dep is in `devDependencies`, is will be externalized as well?

Comment (Timeless0911) [reply packages/core/src/utils/external.ts:45]: `if (externalOptions[type])` do judgement?

File: packages/core/src/utils/external.ts (hunk 3 above)
Comment (9aoy) [reply packages/core/src/utils/external.ts:36]:
```
>  /.a/ instanceof Object => true
>  Object.prototype.toString.call(/.a/) => '[object RegExp]'
```

Comment (fi3ework) [reply packages/core/src/utils/external.ts:36]: Awesome JaVa$cRipT.

File: packages/core/src/utils/external.ts (hunk 4 above)
Comment (9aoy) [reply packages/core/src/utils/external.ts:45]: `devDependencies` will not externalized by default,  so do we need to support `development` configuration item in autoExternal?

```diff
type AutoExternal =
  | boolean
  | {
      dependencies?: boolean;
-     devDependencies?: boolean;
      peerDependencies?: boolean;
    };
```

Comment (fi3ework) [reply packages/core/src/utils/external.ts:45]: I saw that, LGTM, no more 1. question.
If a dep is in `peerDependencies` and `dependencies` (I don't know why user does so), we still need a dedup thought.

**Comment by 9aoy** [comment]
> Can we update examples `externals` config?

The example scenario is not yet supported. When we external `react`, `react/jsx-runtime` will not be externalized.  And I think `autoExternal` should support this feature (next PR).

File: packages/core/src/utils/external.ts (hunk 4 above)
Comment (9aoy) [reply packages/core/src/utils/external.ts:45]: Okay, let's dedup. In fact, duplicate keys have no impact on the function. 😆

Comment (9aoy) [reply packages/core/src/utils/external.ts]: done

**Comment by chenjiahan** [comment]
> When we external react, react/jsx-runtime will not be externalized.

Why is `react/jsx-runtime` not externalized, I thought it should work.

**Comment by Timeless0911** [comment]
https://github.com/web-infra-dev/modern.js/blob/3824a0e819d2444fb220aa96491c2971c998d514/packages/solutions/module-tools/src/config/merge.ts#L91-L105

What Modern.js Module do

**Comment by 9aoy** [comment]
> Why is react/jsx-runtime not externalized, I thought it should work.

done

**Review by Timeless0911 — APPROVED** [review]

**Pull request merged**